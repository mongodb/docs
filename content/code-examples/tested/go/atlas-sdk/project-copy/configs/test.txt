A nested file to test that the copier app is working.