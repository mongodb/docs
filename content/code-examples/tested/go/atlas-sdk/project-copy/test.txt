Test file to verify that the copier app is working. 