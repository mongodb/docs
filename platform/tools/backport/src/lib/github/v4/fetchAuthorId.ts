import gql from 'graphql-tag';
import { apiRequestV4 } from './apiRequestV4';

export interface AuthorIdResponse {
  user: { id: string };
}

export async function fetchAuthorId({
  accessToken,
  author,
  githubApiBaseUrlV4 = 'https://api.github.com/graphql',
}: {
  accessToken: string;
  author: string | null;
  githubApiBaseUrlV4?: string;
}) {
  if (author === null) {
    return null;
  }

  const query = gql`
    query AuthorId($author: String!) {
      user(login: $author) {
        id
      }
    }
  `;

  const res = await apiRequestV4<AuthorIdResponse>({
    githubApiBaseUrlV4,
    accessToken,
    query,
    variables: { author },
  });

  return res.data.data.user.id;
}
