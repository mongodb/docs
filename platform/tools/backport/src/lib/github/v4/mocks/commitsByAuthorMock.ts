import { CommitByAuthorResponse } from '../fetchCommits/fetchCommitsByAuthor';

export const commitsByAuthorMock: CommitByAuthorResponse = {
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                author: {
                  email: 'soren.louv@elastic.co',
                  name: 'SÃ¸ren Louv-Jansen',
                },
                repository: {
                  name: 'kibana',
                  owner: { login: 'elastic' },
                },
                committedDate: '2021-12-24T00:00:00Z',
                sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
                message: 'Add ðŸ‘»',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                author: {
                  email: 'soren.louv@elastic.co',
                  name: 'SÃ¸ren Louv-Jansen',
                },
                repository: {
                  name: 'kibana',
                  owner: { login: 'elastic' },
                },
                committedDate: '2021-12-23T00:00:00Z',
                sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                message: 'Add witch (#85)',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        title: 'Derive Queries Panel',
                        url: 'https://github.com/elastic/kibana/pull/85',
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-b' }],
                        },
                        number: 85,
                        mergeCommit: {
                          remoteConfigHistory: { edges: [] },
                          sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                          message: 'Add witch (#85)',
                        },
                        timelineItems: {
                          edges: [],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                author: {
                  email: 'soren.louv@elastic.co',
                  name: 'SÃ¸ren Louv-Jansen',
                },
                repository: {
                  name: 'kibana',
                  owner: { login: 'elastic' },
                },
                committedDate: '2021-12-22T00:00:00Z',
                sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        title: 'Creating own query',
                        url: 'https://github.com/elastic/kibana/pull/80',
                        labels: {
                          nodes: [{ name: 'v6.3.0' }],
                        },
                        baseRefName: 'master',
                        number: 80,
                        mergeCommit: {
                          remoteConfigHistory: { edges: [] },
                          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                targetPullRequest: {
                                  __typename: 'PullRequest',
                                  targetMergeCommit: {
                                    sha: 'target-merge-commit-sha',
                                    message:
                                      'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                  },
                                  repository: {
                                    name: 'kibana',
                                    owner: { login: 'elastic' },
                                  },
                                  url: 'https://github.com/elastic/kibana/pull/99',
                                  title: 'some title',
                                  state: 'MERGED',
                                  number: 99,
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          targetCommit: {
                                            sha: 'abc',
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                author: {
                  email: 'soren.louv@elastic.co',
                  name: 'SÃ¸ren Louv-Jansen',
                },
                repository: {
                  name: 'kibana',
                  owner: { login: 'elastic' },
                },
                committedDate: '2021-12-21T00:00:00Z',
                sha: '3827bbbaf39914eda4f02f6940189844375fd097',
                message: 'Add backport config',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                author: {
                  email: 'soren.louv@elastic.co',
                  name: 'SÃ¸ren Louv-Jansen',
                },
                repository: {
                  name: 'kibana',
                  owner: { login: 'elastic' },
                },
                committedDate: '2021-12-20T00:00:00Z',
                sha: '5ea0da550ac191029459289d67f99ad7d310812b',
                message: 'Initial commit',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
          ],
        },
      },
    },
  },
};
