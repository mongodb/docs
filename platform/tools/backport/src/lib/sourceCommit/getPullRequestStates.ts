import { keyBy, merge, uniqBy, values } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import { filterNil } from '../../utils/filterEmpty';
import { getFirstLine } from '../github/commitFormatters';
import {
  SourcePullRequestNode,
  SourceCommitWithTargetPullRequest,
  TimelineEdge,
  TimelinePullRequestEdge,
} from './parseSourceCommit';

export type PullRequestState = 'OPEN' | 'CLOSED' | 'MERGED' | 'NOT_CREATED';
type CreatedPullRequest = {
  url: string;
  number: number;
  branch: string;
  state: PullRequestState;
  mergeCommit?: {
    sha: string;
    message: string;
  };
};

type TargetBranchWithLabel = {
  branchLabelMappingKey: string;
  branch: string;
  label: string;
  isSourceBranch: boolean;
};

export type TargetPullRequest =
  | (CreatedPullRequest & Partial<TargetBranchWithLabel>)
  | ((TargetBranchWithLabel & Partial<CreatedPullRequest>) & {
      state: PullRequestState;
    });

export function getSourcePullRequest(
  sourceCommit: SourceCommitWithTargetPullRequest,
): SourcePullRequestNode | undefined {
  return sourceCommit.associatedPullRequests.edges?.[0]?.node;
}

function mergeByKey<T, K>(
  obj1: T[],
  obj2: K[],
  key: string,
): Array<(T & Partial<K>) | (K & Partial<T>)> {
  const merged = merge(keyBy(obj1, key), keyBy(obj2, key));
  const a = values(merged);
  return a;
}

export function getPullRequestStates({
  sourceCommit,
  branchLabelMapping,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  branchLabelMapping: ValidConfigOptions['branchLabelMapping'];
}): TargetPullRequest[] {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);

  // if there is no source pull request the commit was pushed directly to the source branch
  // in that case there will be no labels, and thus not possible to deduce the expected target branches
  if (!sourcePullRequest) {
    return [];
  }

  const createdTargetPullRequests = getCreatedTargetPullRequests(sourceCommit);

  // if there's no `branchLabelMapping`, it's not possible to determine the missing target branches
  if (!branchLabelMapping) {
    return createdTargetPullRequests;
  }

  const targetBranchesFromLabels = getTargetBranchesFromLabels(
    sourcePullRequest,
    branchLabelMapping,
  );

  return mergeByKey(
    targetBranchesFromLabels,
    createdTargetPullRequests,
    'branch',
  ).map((res) => {
    if (res.state) {
      return { ...res, state: res.state };
    }

    // MERGED (source branch)
    if (res.isSourceBranch) {
      return {
        ...res,
        state: 'MERGED' as const,
        url: sourcePullRequest.url,
        number: sourcePullRequest.number,
        mergeCommit: sourcePullRequest.mergeCommit
          ? {
              message: sourcePullRequest.mergeCommit.message,
              sha: sourcePullRequest.mergeCommit.sha,
            }
          : undefined,
      };
    }

    // NOT_CREATED
    return { ...res, state: 'NOT_CREATED' as const };
  });
}

function getCreatedTargetPullRequests(
  sourceCommit: SourceCommitWithTargetPullRequest,
): CreatedPullRequest[] {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);

  if (!sourcePullRequest) {
    return [];
  }

  const sourceCommitMessage = getFirstLine(sourceCommit.message);
  return sourcePullRequest.timelineItems.edges
    .filter(filterNil)
    .filter(filterPullRequests)
    .filter((item) => {
      const { targetPullRequest } = item.node;

      // ignore closed PRs
      if (targetPullRequest.state === 'CLOSED') {
        return false;
      }

      // at least one of the commits in `targetPullRequest` should match the merge commit from the source pull request
      const didCommitMatch = targetPullRequest.commits.edges.some(
        (commitEdge) => {
          const { targetCommit } = commitEdge.node;

          const matchingRepoName =
            sourceCommit.repository.name === targetPullRequest.repository.name;

          const matchingRepoOwner =
            sourceCommit.repository.owner.login ===
            targetPullRequest.repository.owner.login;

          const targetCommitMessage = getFirstLine(targetCommit.message);

          const matchingMessage = targetCommitMessage === sourceCommitMessage;
          return matchingRepoName && matchingRepoOwner && matchingMessage;
        },
      );

      const titleIncludesMessage =
        targetPullRequest.title.includes(sourceCommitMessage);

      const titleIncludesNumber = targetPullRequest.title.includes(
        sourcePullRequest.number.toString(),
      );

      return didCommitMatch || (titleIncludesMessage && titleIncludesNumber);
    })
    .map((item) => {
      const { targetPullRequest } = item.node;
      return {
        url: targetPullRequest.url,
        number: targetPullRequest.number,
        branch: targetPullRequest.baseRefName,
        state: targetPullRequest.state,
        mergeCommit: targetPullRequest.targetMergeCommit
          ? {
              sha: targetPullRequest.targetMergeCommit.sha,
              message: targetPullRequest.targetMergeCommit.message,
            }
          : undefined,
      };
    });
}

// narrow TimelineEdge to TimelinePullRequestEdge
function filterPullRequests(
  item: TimelineEdge,
): item is TimelinePullRequestEdge {
  const { targetPullRequest } = item.node;
  return targetPullRequest.__typename === 'PullRequest';
}

function getTargetBranchesFromLabels(
  sourcePullRequest: SourcePullRequestNode,
  branchLabelMapping: NonNullable<ValidConfigOptions['branchLabelMapping']>,
): TargetBranchWithLabel[] {
  const targetBranchesFromLabels = sourcePullRequest.labels.nodes
    .map((label) => label.name)
    .map((label) => {
      const res = getTargetBranchFromLabel({ branchLabelMapping, label });
      if (res) {
        const { branchLabelMappingKey, targetBranch } = res;
        const isSourceBranch = targetBranch === sourcePullRequest.baseRefName;
        return {
          branch: targetBranch,
          label,
          branchLabelMappingKey,
          isSourceBranch,
        };
      }
    })
    .filter(filterNil);

  return uniqBy(targetBranchesFromLabels, ({ branch }) => branch);
}

export function getTargetBranchFromLabel({
  branchLabelMapping,
  label,
}: {
  branchLabelMapping: NonNullable<ValidConfigOptions['branchLabelMapping']>;
  label: string;
}) {
  // only get first match
  const result = Object.entries(branchLabelMapping).find(([key]) => {
    const regex = new RegExp(key);
    const isMatch = label.match(regex) !== null;
    return isMatch;
  });

  if (result) {
    const [branchLabelMappingKey, branchLabelMappingValue] = result;
    const regex = new RegExp(branchLabelMappingKey);
    const targetBranch = label.replace(regex, branchLabelMappingValue);

    if (targetBranch) {
      return { targetBranch, branchLabelMappingKey };
    }
  }
}
