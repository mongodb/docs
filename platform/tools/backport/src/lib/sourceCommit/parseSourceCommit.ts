import gql from 'graphql-tag';
import { differenceBy } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import {
  RemoteConfigHistory,
  RemoteConfigHistoryFragment,
  parseRemoteConfigFile,
} from '../remoteConfig';
import {
  TargetPullRequest,
  getPullRequestStates,
  getSourcePullRequest,
} from './getPullRequestStates';

export interface Commit {
  author: SourceCommitWithTargetPullRequest['author'];
  sourceCommit: {
    committedDate: string;
    message: string;
    sha: string;
    branchLabelMapping: ValidConfigOptions['branchLabelMapping'];
  };
  sourcePullRequest?: {
    title: string;
    labels: string[];
    number: number;
    url: string;
    mergeCommit?: {
      message: string;
      sha: string;
    };
  };
  sourceBranch: string;
  suggestedTargetBranches: string[];
  // suggestedTargetDirectories: string[]; // TODO: Feature
  targetPullRequestStates: TargetPullRequest[];
}

export interface SourcePullRequestNode {
  title: string;
  baseRefName: string;
  url: string;
  number: number;
  labels: {
    nodes: {
      name: string;
    }[];
  };
  mergeCommit?: {
    remoteConfigHistory: RemoteConfigHistory['remoteConfigHistory'];
    sha: string;
    message: string;
  };
  timelineItems: {
    edges: TimelineEdge[];
  };
}

export type TimelineEdge = TimelinePullRequestEdge | TimelineIssueEdge;

export interface TimelinePullRequestEdge {
  node: {
    targetPullRequest: {
      __typename: 'PullRequest';
      url: string;
      title: string;
      state: 'OPEN' | 'CLOSED' | 'MERGED';
      baseRefName: string;
      number: number;

      targetMergeCommit: {
        sha: string;
        message: string;
      } | null;

      repository: {
        name: string;
        owner: {
          login: string;
        };
      };

      commits: {
        edges: Array<{
          node: { targetCommit: { message: string; sha: string } };
        }>;
      };
    };
  };
}

interface TimelineIssueEdge {
  node: { targetPullRequest: { __typename: 'Issue' } };
}

export type SourceCommitWithTargetPullRequest = {
  repository: {
    name: string;
    owner: { login: string };
  };
  sha: string;
  message: string;
  committedDate: string;
  author: {
    name: string;
    email: string;
  };

  associatedPullRequests: {
    edges: { node: SourcePullRequestNode }[] | null;
  };
};

function getSuggestedTargetBranches(
  sourceCommit: SourceCommitWithTargetPullRequest,
  targetPullRequestStates: TargetPullRequest[],
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'],
) {
  const missingPrs = getPullRequestStates({
    sourceCommit,
    branchLabelMapping,
  }).filter((pr) => pr.state === 'NOT_CREATED' || pr.state === 'CLOSED');

  const mergedPrs = targetPullRequestStates.filter(
    (pr) => pr.state === 'MERGED',
  );

  return differenceBy(missingPrs, mergedPrs, (pr) => pr.label).map(
    (pr) => pr.branch,
  );
}

export function parseSourceCommit({
  sourceCommit,
  options,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  options: {
    branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
    sourceBranch: string;
  };
}): Commit {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);
  const sourceCommitBranchLabelMapping =
    getSourceCommitBranchLabelMapping(sourceCommit);

  const currentBranchLabelMapping = options.branchLabelMapping;

  const targetPullRequestStates = getPullRequestStates({
    sourceCommit,
    branchLabelMapping:
      sourceCommitBranchLabelMapping ?? currentBranchLabelMapping,
  });

  const suggestedTargetBranches = getSuggestedTargetBranches(
    sourceCommit,
    targetPullRequestStates,
    currentBranchLabelMapping,
  );

  return {
    author: sourceCommit.author,
    sourceCommit: {
      committedDate: sourceCommit.committedDate,
      message: sourceCommit.message,
      sha: sourceCommit.sha,
      branchLabelMapping: sourceCommitBranchLabelMapping,
    },
    sourcePullRequest: sourcePullRequest
      ? {
          labels: sourcePullRequest.labels.nodes.map((label) => label.name),
          title: sourcePullRequest.title,
          number: sourcePullRequest.number,
          url: sourcePullRequest.url,
          mergeCommit: sourcePullRequest.mergeCommit
            ? {
                message: sourcePullRequest.mergeCommit.message,
                sha: sourcePullRequest.mergeCommit.sha,
              }
            : undefined,
        }
      : undefined,
    sourceBranch: sourcePullRequest?.baseRefName ?? options.sourceBranch,
    suggestedTargetBranches,
    targetPullRequestStates: targetPullRequestStates,
  };
}

export const SourceCommitWithTargetPullRequestFragment = gql`
  fragment SourceCommitWithTargetPullRequestFragment on Commit {
    # Source Commit
    repository {
      name
      owner {
        login
      }
    }
    sha: oid
    message
    committedDate

    author {
      name
      email
    }

    # Source pull request: PR where source commit was merged in
    associatedPullRequests(first: 1) {
      edges {
        node {
          title
          url
          number
          labels(first: 50) {
            nodes {
              name
            }
          }
          baseRefName

          # source merge commit (the commit that actually went into the source branch)
          mergeCommit {
            ...RemoteConfigHistoryFragment
            sha: oid
            message
          }

          # (possible) backport pull requests referenced in the source pull request
          timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
            edges {
              node {
                ... on CrossReferencedEvent {
                  targetPullRequest: source {
                    __typename

                    # Target PRs (backport PRs)
                    ... on PullRequest {
                      # target merge commit: the backport commit that was merged into the target branch
                      targetMergeCommit: mergeCommit {
                        sha: oid
                        message
                      }
                      repository {
                        name
                        owner {
                          login
                        }
                      }
                      url
                      title
                      state
                      baseRefName
                      number
                      commits(first: 20) {
                        edges {
                          node {
                            targetCommit: commit {
                              message
                              sha: oid
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  ${RemoteConfigHistoryFragment}
`;

function getSourceCommitBranchLabelMapping(
  sourceCommit: SourceCommitWithTargetPullRequest,
): ValidConfigOptions['branchLabelMapping'] {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);

  const remoteConfig =
    sourcePullRequest?.mergeCommit?.remoteConfigHistory.edges?.[0]
      ?.remoteConfig;

  if (remoteConfig) {
    return parseRemoteConfigFile(remoteConfig)?.branchLabelMapping;
  }
}
