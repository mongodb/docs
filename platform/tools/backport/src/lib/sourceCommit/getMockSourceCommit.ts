import { SourceCommitWithTargetPullRequest } from './parseSourceCommit';

export function getMockSourceCommit({
  sourceCommit,
  sourcePullRequest,
  timelineItems = [],
}: {
  sourceCommit: {
    remoteConfig?: {
      branchLabelMapping: Record<string, string>;
      committedDate: string;
    };
    sha?: string;
    message: string;
    commitedDate?: string;
  };
  sourcePullRequest: {
    title?: string;
    number: number;
    labels?: string[];
    sourceBranch?: string;
  } | null;
  timelineItems?: Array<{
    state: 'OPEN' | 'CLOSED' | 'MERGED';
    targetBranch: string;
    title?: string;
    number: number;
    commitMessages: string[];
    repoName?: string;
    repoOwner?: string;
  }>;
}): SourceCommitWithTargetPullRequest {
  const defaultTargetPullRequestTitle =
    'DO NOT USE: Please specify a title in test!!!';

  const defaultSourceCommitSha = 'DO NOT USE: please specify a sha in test!!!';

  const baseMockCommit: SourceCommitWithTargetPullRequest = {
    author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
    repository: {
      name: 'kibana',
      owner: { login: 'elastic' },
    },
    committedDate: sourceCommit.commitedDate ?? '2021-12-22T00:00:00Z',
    sha: sourceCommit.sha ?? defaultSourceCommitSha,
    message: sourceCommit.message,
    associatedPullRequests: { edges: null },
  };

  if (!sourcePullRequest) {
    return baseMockCommit;
  }

  const remoteConfigHistory = sourceCommit.remoteConfig
    ? {
        edges: [
          {
            remoteConfig: {
              committedDate: sourceCommit.remoteConfig.committedDate,
              file: {
                object: {
                  text: JSON.stringify({
                    branchLabelMapping:
                      sourceCommit.remoteConfig.branchLabelMapping,
                  }),
                },
              },
            },
          },
        ],
      }
    : { edges: [] };

  return {
    ...baseMockCommit,
    associatedPullRequests: {
      edges: [
        {
          node: {
            mergeCommit: {
              remoteConfigHistory,
              sha: sourceCommit.sha ?? defaultSourceCommitSha,
              message: sourceCommit.message,
            },
            url: `https://github.com/elastic/kibana/pull/${sourcePullRequest.number}`,
            title: sourcePullRequest.title ?? defaultTargetPullRequestTitle,
            labels: {
              nodes: (sourcePullRequest.labels ?? []).map((name) => ({
                name,
              })),
            },
            baseRefName:
              sourcePullRequest.sourceBranch ??
              'source-branch-from-associated-pull-request',
            number: sourcePullRequest.number,
            timelineItems: {
              edges: timelineItems.map((timelineItem) => {
                return {
                  node: {
                    targetPullRequest: {
                      __typename: 'PullRequest',
                      url: `https://github.com/elastic/kibana/pull/${timelineItem.number}`,
                      title:
                        timelineItem.title ?? defaultTargetPullRequestTitle,
                      number: timelineItem.number,
                      state: timelineItem.state,
                      baseRefName: timelineItem.targetBranch,

                      targetMergeCommit:
                        timelineItem.state === 'MERGED'
                          ? {
                              message: timelineItem.commitMessages[0],
                              sha: 'target-merge-commit-sha',
                            }
                          : null,

                      repository: {
                        name: timelineItem.repoName ?? 'kibana',
                        owner: {
                          login: timelineItem.repoOwner ?? 'elastic',
                        },
                      },
                      commits: {
                        edges: timelineItem.commitMessages.map((message) => ({
                          node: {
                            targetCommit: {
                              sha: 'abc',
                              message: message,
                            },
                          },
                        })),
                      },
                    },
                  },
                };
              }),
            },
          },
        },
      ],
    },
  };
}
