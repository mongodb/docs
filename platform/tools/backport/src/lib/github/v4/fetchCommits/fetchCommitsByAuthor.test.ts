import nock from 'nock';
import { mockGqlRequest } from '../../../../test/nockHelpers';
import { Commit } from '../../../sourceCommit/parseSourceCommit';
import { AuthorIdResponse } from '../fetchAuthorId';
import { commitsByAuthorMock } from '../mocks/commitsByAuthorMock';
import {
  CommitByAuthorResponse,
  fetchCommitsByAuthor,
} from './fetchCommitsByAuthor';

const defaultOptions = {
  accessToken: 'myAccessToken',
  author: 'sorenlouv',
  githubApiBaseUrlV4: 'http://localhost/graphql',
  maxNumber: 10,
  repoName: 'kibana',
  repoOwner: 'elastic',
  sourceBranch: 'source-branch-from-options',
  dateSince: null,
  dateUntil: null,
};

const authorIdMockData = { user: { id: 'myUserId' } } as const;

describe('fetchCommitsByAuthor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    nock.cleanAll();
  });

  describe('when commit has an associated pull request', () => {
    let res: Commit[];
    let authorIdCalls: ReturnType<typeof mockGqlRequest>;
    let commitsByAuthorCalls: ReturnType<typeof mockGqlRequest>;

    beforeEach(async () => {
      authorIdCalls = mockGqlRequest<AuthorIdResponse>({
        name: 'AuthorId',
        statusCode: 200,
        body: { data: authorIdMockData },
      });

      commitsByAuthorCalls = mockGqlRequest<CommitByAuthorResponse>({
        name: 'CommitsByAuthor',
        statusCode: 200,
        body: { data: commitsByAuthorMock },
      });

      res = await fetchCommitsByAuthor(defaultOptions);
    });

    it('should return a list of commits with pullNumber and existing backports', () => {
      const expectedCommits: Commit[] = [
        {
          author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: undefined,
            committedDate: '2021-12-24T00:00:00Z',
            sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
            message: 'Add ðŸ‘»',
          },
          targetPullRequestStates: [],
          sourceBranch: 'source-branch-from-options',
        },
        {
          author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: undefined,
            committedDate: '2021-12-23T00:00:00Z',
            sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
            message: 'Add witch (#85)',
          },
          sourcePullRequest: {
            title: 'Derive Queries Panel',
            labels: ['my-label-b'],
            number: 85,
            url: 'https://github.com/elastic/kibana/pull/85',
            mergeCommit: {
              sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
              message: 'Add witch (#85)',
            },
          },
          targetPullRequestStates: [],
          sourceBranch: 'master',
        },
        {
          author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: undefined,
            committedDate: '2021-12-22T00:00:00Z',
            sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
            message:
              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
          },
          sourcePullRequest: {
            title: 'Creating own query',
            labels: ['v6.3.0'],
            number: 80,
            url: 'https://github.com/elastic/kibana/pull/80',
            mergeCommit: {
              sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
              message:
                'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
            },
          },
          targetPullRequestStates: [
            {
              branch: '6.3',
              state: 'MERGED',
              number: 99,
              url: 'https://github.com/elastic/kibana/pull/99',
              mergeCommit: {
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                sha: 'target-merge-commit-sha',
              },
            },
          ],
          sourceBranch: 'master',
        },
        {
          author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: undefined,
            committedDate: '2021-12-21T00:00:00Z',
            sha: '3827bbbaf39914eda4f02f6940189844375fd097',
            message: 'Add backport config',
          },
          targetPullRequestStates: [],
          sourceBranch: 'source-branch-from-options',
        },
        {
          author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
          suggestedTargetBranches: [],
          sourceCommit: {
            branchLabelMapping: undefined,
            committedDate: '2021-12-20T00:00:00Z',
            sha: '5ea0da550ac191029459289d67f99ad7d310812b',
            message: 'Initial commit',
          },
          targetPullRequestStates: [],
          sourceBranch: 'source-branch-from-options',
        },
      ];
      expect(res).toEqual(expectedCommits);
    });

    it('should call with correct args to fetch author id', () => {
      expect(authorIdCalls).toMatchSnapshot();
    });

    it('should call with correct args to fetch commits', () => {
      expect(commitsByAuthorCalls).toMatchSnapshot();
    });
  });

  describe('when a custom github api hostname is supplied', () => {
    it('should be used in gql requests', async () => {
      const authorIdCalls = mockGqlRequest<AuthorIdResponse>({
        name: 'AuthorId',
        statusCode: 200,
        body: { data: authorIdMockData },
        apiBaseUrl: 'http://localhost/my-custom-api',
      });

      const commitsByAuthorCalls = mockGqlRequest<CommitByAuthorResponse>({
        name: 'CommitsByAuthor',
        statusCode: 200,
        body: { data: commitsByAuthorMock },
        apiBaseUrl: 'http://localhost/my-custom-api',
      });

      await fetchCommitsByAuthor({
        ...defaultOptions,
        githubApiBaseUrlV4: 'http://localhost/my-custom-api',
      });

      expect(authorIdCalls.length).toBe(1);
      expect(commitsByAuthorCalls.length).toBe(1);
    });
  });
});
