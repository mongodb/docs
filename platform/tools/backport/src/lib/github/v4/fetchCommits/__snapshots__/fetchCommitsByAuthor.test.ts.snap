// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch author id 1`] = `
[
  {
    "query": "query AuthorId($author: String!) {
  user(login: $author) {
    id
  }
}",
    "variables": {
      "author": "sorenlouv",
    },
  },
]
`;

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch commits 1`] = `
[
  {
    "query": "query CommitsByAuthor($authorId: ID, $commitPath: String, $dateSince: GitTimestamp, $dateUntil: GitTimestamp, $maxNumber: Int!, $repoName: String!, $repoOwner: String!, $sourceBranch: String!) {
  repository(owner: $repoOwner, name: $repoName) {
    ref(qualifiedName: $sourceBranch) {
      target {
        ... on Commit {
          history(
            first: $maxNumber
            author: {id: $authorId}
            path: $commitPath
            since: $dateSince
            until: $dateUntil
          ) {
            edges {
              node {
                ...SourceCommitWithTargetPullRequestFragment
              }
            }
          }
        }
      }
    }
  }
}

fragment SourceCommitWithTargetPullRequestFragment on Commit {
  repository {
    name
    owner {
      login
    }
  }
  sha: oid
  message
  committedDate
  author {
    name
    email
  }
  associatedPullRequests(first: 1) {
    edges {
      node {
        title
        url
        number
        labels(first: 50) {
          nodes {
            name
          }
        }
        baseRefName
        mergeCommit {
          ...RemoteConfigHistoryFragment
          sha: oid
          message
        }
        timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
          edges {
            node {
              ... on CrossReferencedEvent {
                targetPullRequest: source {
                  __typename
                  ... on PullRequest {
                    targetMergeCommit: mergeCommit {
                      sha: oid
                      message
                    }
                    repository {
                      name
                      owner {
                        login
                      }
                    }
                    url
                    title
                    state
                    baseRefName
                    number
                    commits(first: 20) {
                      edges {
                        node {
                          targetCommit: commit {
                            message
                            sha: oid
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

fragment RemoteConfigHistoryFragment on Commit {
  remoteConfigHistory: history(first: 1, path: ".backportrc.json") {
    edges {
      remoteConfig: node {
        committedDate
        file(path: ".backportrc.json") {
          ... on TreeEntry {
            object {
              ... on Blob {
                text
              }
            }
          }
        }
      }
    }
  }
}",
    "variables": {
      "authorId": "myUserId",
      "commitPath": null,
      "dateSince": null,
      "dateUntil": null,
      "maxNumber": 10,
      "repoName": "kibana",
      "repoOwner": "elastic",
      "sourceBranch": "source-branch-from-options",
    },
  },
]
`;
